tower-http/src/request_id.rs

```diff
@@ -0,0 +1,575 @@
+    pub fn new(header_name: HeaderName, make_request_id: M) -> Self
+    where
+        M: MakeRequestId,
+    {
+        SetRequestIdLayer {
+            header_name,
+            make_request_id,
+        }
+    }
+
+    /// Create a new `SetRequestIdLayer` that uses `x-request-id` as the header name.
+    pub fn x_request_id(make_request_id: M) -> Self
+    where
+        M: MakeRequestId,
+    {
+        SetRequestIdLayer {
+            header_name: HeaderName::from_static(X_REQUEST_ID),
+            make_request_id,
+        }
+    }
+}
+
+impl<S, M> Layer<S> for SetRequestIdLayer<M>
+where
+    M: Clone + MakeRequestId,
+{
+    type Service = SetRequestId<S, M>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        SetRequestId::new(
+            inner,
+            self.header_name.clone(),
+            self.make_request_id.clone(),
+        )
+    }
+}
+
+/// Set request id headers and extensions on requests.
+///
+/// See the [module docs](self) for an example.
+///
+/// If [`MakeRequestId::make_request_id`] returns `Some(_)` and the request doesn't already have a
+/// header with the same name, then the header will be inserted.
+///
+/// Additionally [`RequestId`] will be inserted into [`Request::extensions`] so other
+/// services can access it.
+#[derive(Debug, Clone)]
+pub struct SetRequestId<S, M> {
+    inner: S,
+    header_name: HeaderName,
+    make_request_id: M,
+}
+
+impl<S, M> SetRequestId<S, M> {
+    /// Create a new `SetRequestId`.
+    pub fn new(inner: S, header_name: HeaderName, make_request_id: M) -> Self
+    where
+        M: MakeRequestId,
+    {
+        Self {
+            inner,
+            header_name,
+            make_request_id,
+        }
+    }
+
+    /// Create a new `SetRequestId` that uses `x-request-id` as the header name.
+    pub fn x_request_id(inner: S, make_request_id: M) -> Self
+    where
+        M: MakeRequestId,
+    {
+        Self::new(
+            inner,
+            HeaderName::from_static(X_REQUEST_ID),
+            make_request_id,
+        )
+    }
+
+    define_inner_service_accessors!();
+
+    /// Returns a new [`Layer`] that wraps services with a `SetRequestId` middleware.
+    pub fn layer(header_name: HeaderName, make_request_id: M) -> SetRequestIdLayer<M>
+    where
+        M: MakeRequestId,
+    {
+        SetRequestIdLayer::new(header_name, make_request_id)
+    }
+}
+
+impl<S, M, ReqBody, ResBody> Service<Request<ReqBody>> for SetRequestId<S, M>
+where
+    S: Service<Request<ReqBody>, Response = Response<ResBody>>,
+    M: MakeRequestId,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = S::Future;
+
+    #[inline]
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, mut req: Request<ReqBody>) -> Self::Future {
+        if let Some(request_id) = req.headers().get(&self.header_name).cloned() {
```

Comment:
`clone` should be done inside branch when `req.extensions().get::<RequestId>().is_none()`